crates/block-producer/src/batch_builder/batch.rs

```diff
@@ -26,80 +32,122 @@ pub type BatchId = Blake3Digest<32>;
 pub struct TransactionBatch {
     id: BatchId,
     updated_accounts: Vec<(TransactionId, TxAccountUpdate)>,
-    unauthenticated_input_notes: BTreeSet<NoteId>,
-    produced_nullifiers: Vec<Nullifier>,
+    input_notes: Vec<InputNoteCommitment>,
     output_notes_smt: BatchNoteTree,
     output_notes: Vec<OutputNote>,
 }
 
 impl TransactionBatch {
-    // CONSTRUCTOR
+    // CONSTRUCTORS
     // --------------------------------------------------------------------------------------------
+
     /// Returns a new [TransactionBatch] instantiated from the provided vector of proven
-    /// transactions.
+    /// transactions. If a map of unauthenticated notes found in the store is provided, it is used
+    /// for transforming unauthenticated notes into authenticated notes.
     ///
     /// # Errors
     /// Returns an error if:
-    /// - The number of created notes across all transactions exceeds 4096.
+    /// - The number of output notes across all transactions exceeds 4096.
+    /// - There are duplicated output notes or unauthenticated notes found across all transactions
+    ///   in the batch.
+    /// - Hashes for corresponding input notes and output notes don't match.
     ///
     /// TODO: enforce limit on the number of created nullifiers.
     #[instrument(target = "miden-block-producer", name = "new_batch", skip_all, err)]
-    pub fn new(txs: Vec<ProvenTransaction>) -> Result<Self, BuildBatchError> {
+    pub fn new(
+        txs: Vec<ProvenTransaction>,
+        found_unauthenticated_notes: Option<BTreeMap<NoteId, MerklePath>>,
+    ) -> Result<Self, BuildBatchError> {
         let id = Self::compute_id(&txs);
 
+        // Populate batch output notes and updated accounts.
         let mut updated_accounts = vec![];
-        let mut produced_nullifiers = vec![];
+        let mut output_notes = vec![];
+        let mut output_note_index = BTreeMap::new();
```

Comment:
I think it could improve the readability of the code quite a bit if we combined these two variables into a dedicated struct - e.g., something like `OutputNoteTracker`. We could then implement various convenience methods on this struct. For example:
```rust
impl OutputNoteTracker {
    pub fn add_notes_from_tx(&mut self, tx: &ProvenTransaction) -> Result<(), BuildBatchError> {
         todo!()
    }

    pub fn remove_note(&mut self, note_header: &NoteHeader) -> Result<bool, BuildBatchError> {
        todo!()
    }

    pub fn into_notes(&self) -> Vec<OutputNote> {
        todo!()
    }
}
```